### ISSUE TITLE
Is there anything example about How to apply model saved by distribution tensorflow in tensorflow serving?

### ISSUE THREAD

**Issue by terryKing1992** [tensorflow/serving#363]
I have tried three saving model methods, eg: `exporter` 、`save_model_builder` and `tf.train.Supervisor(logdir=XXX)`, and i met series of issues on each method, could you please provide some tensorflow distribution examples on saving model for serving? thanks a lot

**Comment by terryKing1992** [comment]
anyone else can help me? please

**Comment by kirilg** [comment]
We have a number of examples of exporting using the SavedModel format in our examples directory:
https://github.com/tensorflow/serving/blob/master/tensorflow_serving/example/inception_saved_model.py
https://github.com/tensorflow/serving/blob/master/tensorflow_serving/example/mnist_saved_model.py

You can also take a look at this simple model that we use throughout our tests: https://github.com/tensorflow/tensorflow/blob/799e31f3840c21322e380e1ec6e5bacb95d016fa/tensorflow/examples/saved_model/saved_model_half_plus_two.py#L79

Please also check out the documentation [here](https://tensorflow.github.io/serving/serving_basic) which covers basic exporting of a TF model. Let us know if the documentation is not clear or if you encounter any issues (with more details about what the error was).

**Comment by terryKing1992** [comment]
yeah, this example can work very well on standalone machine, but when i use Distribution tensorflow to save the model, there is an error` raise RuntimeError("Graph is finalized and cannot be modified.")`.

and my distribution code like this:

`
# _*_ coding:utf-8 _*_
    import math
    import time
    import tensorflow as tf
    from tensorflow.contrib.session_bundle import exporter
    from tensorflow.python.saved_model import builder as saved_model_builder
    from tensorflow.python.saved_model import signature_constants
    from tensorflow.python.saved_model import signature_def_utils
    from tensorflow.python.saved_model import tag_constants
    from tensorflow.python.saved_model import utils
    from tensorflow.python.util import compat
    
    from tensorflow.examples.tutorials.mnist import input_data
    
    tf.app.flags.DEFINE_string("ps_hosts", "", "Comma-separated list of hostname:port pairs")
    
    tf.app.flags.DEFINE_string("worker_hosts", "", "Comma-separated list of hostname:port pairs")
    
    tf.app.flags.DEFINE_string("job_name", "", "One of 'ps', 'worker'")
    
    tf.app.flags.DEFINE_integer("task_index", 0, "Index of task within the job")
    
    tf.app.flags.DEFINE_integer("hidden_units", 100, "Number of units in the hidden layer of the NN")
    
    tf.app.flags.DEFINE_string("data_dir", "MNIST_data", "Directory for storing mnist data")
    
    tf.app.flags.DEFINE_integer("batch_size", 100, "Training batch size")
    
    FLAGS = tf.app.flags.FLAGS
    
    IMAGE_PIXELS = 28
    
    
    def main(_):
        ps_hosts = FLAGS.ps_hosts.split(",")
    
        worker_hosts = FLAGS.worker_hosts.split(",")
        cluster = tf.train.ClusterSpec({"ps": ps_hosts, "worker": worker_hosts})
        server = tf.train.Server(cluster, job_name=FLAGS.job_name, task_index=FLAGS.task_index)
    
        if FLAGS.job_name == "ps":
            server.join()
    
        elif FLAGS.job_name == "worker":
            start_time = time.time()
            with tf.device(
                    tf.train.replica_device_setter(worker_device="/job:worker/task:%d" % FLAGS.task_index,
                                                   cluster=cluster)):
                hid_w = tf.Variable(
                    tf.truncated_normal([IMAGE_PIXELS * IMAGE_PIXELS, FLAGS.hidden_units], stddev=1.0 / IMAGE_PIXELS),
                    name="hid_w")
    
                hid_b = tf.Variable(tf.zeros([FLAGS.hidden_units]), name="hid_b")
    
                sm_w = tf.Variable(tf.truncated_normal([FLAGS.hidden_units, 10], stddev=1.0 / math.sqrt(FLAGS.hidden_units)),
                                   name="sm_w")
    
                sm_b = tf.Variable(tf.zeros([10]), name="sm_b")
    
                x = tf.placeholder(tf.float32, [None, IMAGE_PIXELS * IMAGE_PIXELS])
    
                y_ = tf.placeholder(tf.float32, [None, 10])
    
                hid_lin = tf.nn.xw_plus_b(x, hid_w, hid_b)
    
                hid = tf.nn.relu(hid_lin)
    
                y = tf.nn.softmax(tf.nn.xw_plus_b(hid, sm_w, sm_b))
    
                loss = -tf.reduce_sum(y_ * tf.log(tf.clip_by_value(y, 1e-10, 1.0)))
    
                global_step = tf.Variable(0)
    
                train_op = tf.train.AdagradOptimizer(0.01).minimize(loss, global_step=global_step)
    
                correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
                accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
    
                saver = tf.train.Saver()
    
                summary_op = tf.summary.merge_all()
    
                init_op = tf.global_variables_initializer()
    
                sv = tf.train.Supervisor(is_chief=(FLAGS.task_index == 0), logdir="train_logs", init_op=init_op,
                                         summary_op=summary_op, saver=saver, global_step=global_step, save_model_secs=600)
                mnist = input_data.read_data_sets(FLAGS.data_dir, one_hot=True)
                with sv.managed_session(server.target) as sess:
                    step = 0
    
                    while not sv.should_stop() and step < 1000:
                        batch_xs, batch_ys = mnist.train.next_batch(FLAGS.batch_size)
                        train_feed = {x: batch_xs, y_: batch_ys}
    
                        _, step = sess.run([train_op, global_step], feed_dict=train_feed)
    
                        if step % 100 == 0:
                            print("global step: {}, accuracy:{}".format(step, sess.run(accuracy,
                                                                                       feed_dict={x: mnist.test.images,
                                                                                                  y_: mnist.test.labels})))
    
                    if sv.is_chief:
                        classification_inputs = utils.build_tensor_info(x)
                        classification_outputs_classes = utils.build_tensor_info(y)
                        classification_signature = signature_def_utils.build_signature_def(
                            inputs={signature_constants.CLASSIFY_INPUTS: classification_inputs},
                            outputs={
                                signature_constants.CLASSIFY_OUTPUT_CLASSES: classification_outputs_classes},
                            method_name=signature_constants.CLASSIFY_METHOD_NAME)
    
                        tensor_info_x = utils.build_tensor_info(x)
                        tensor_info_y = utils.build_tensor_info(y)
    
                        prediction_signature = signature_def_utils.build_signature_def(
                            inputs={'input': tensor_info_x},
                            outputs={'output': tensor_info_y},
                            method_name=signature_constants.PREDICT_METHOD_NAME)
    
                        builder = saved_model_builder.SavedModelBuilder("mnist_model/1")
                        legacy_init_op = tf.group(tf.tables_initializer(), name='legacy_init_op')
    
                        builder.add_meta_graph_and_variables(copy_sess,
                            [tag_constants.SERVING],
                            signature_def_map={
                                'predict_images':
                                    prediction_signature,
                                signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY:
                                    classification_signature,
                            },
                            clear_devices=False,
                            legacy_init_op=legacy_init_op)
    
                        builder.save()
    
        end_time = time.time()
        print('waste time:{}'.format(end_time - start_time))
        sv.stop()
    
    
    if __name__ == "__main__":
        tf.app.run()
`
how can i fix this issue？ perhaps i should new this issue in tensorflow repository. but please

**Comment by kirilg** [comment]
Seems to be the same issue as https://github.com/tensorflow/tensorflow/issues/5439, please give this a try.

**Comment by terryKing1992** [comment]
@kirilg yeah, these two issues have something similarities, but they  are not the same issue. using `exporter` can work well on exporting model. but the `SavedModelBuilder` can't work as `exporter`

**Comment by terryKing1992** [comment]
I Just use some tricks to sovle the issue I mentioned before. before I call `add_meta_graph_and_variables`, and I call this function `sess.graph._unsafe_unfinalize()` to unfinalize the graph, and then call `sess.graph.finalize()` to finalize it. the code like this:

               ```
               if sv.is_chief:
                    sess.graph._unsafe_unfinalize()
                    classification_inputs = utils.build_tensor_info(x)
                    classification_outputs_classes = utils.build_tensor_info(y)

                    classification_signature = signature_def_utils.build_signature_def(
                        inputs={signature_constants.CLASSIFY_INPUTS: classification_inputs},
                        outputs={
                            signature_constants.CLASSIFY_OUTPUT_CLASSES: classification_outputs_classes,
                            signature_constants.CLASSIFY_OUTPUT_SCORES: classification_outputs_classes
                        },
                        method_name=signature_constants.CLASSIFY_METHOD_NAME)

                    tensor_info_x = utils.build_tensor_info(x)
                    tensor_info_y = utils.build_tensor_info(y)

                    prediction_signature = signature_def_utils.build_signature_def(
                        inputs={'images': tensor_info_x},
                        outputs={'scores': tensor_info_y},
                        method_name=signature_constants.PREDICT_METHOD_NAME)

                    builder = saved_model_builder.SavedModelBuilder(logdir)
                    legacy_init_op = tf.group(tf.tables_initializer(), name='legacy_init_op')

                    builder.add_meta_graph_and_variables(sess,
                                                         [tag_constants.SERVING],
                                                         signature_def_map={
                                                             'predict_images':
                                                                 prediction_signature,
                                                             signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY:
                                                                 classification_signature,
                                                         },
                                                         clear_devices=True,
                                                         legacy_init_op=legacy_init_op)
                    sess.graph.finalize()
                    builder.save()
```

**Comment by 0uMuMu0** [comment]
thank you very much!!!
I also met this problem.

**Comment by Xingskcs** [comment]
thank you very much!!!
I also met this problem.

**Comment by zhanwenchen** [comment]
What is the purpose of `utils.build_tensor_info`/`signature_def_map` and how would you adapt this snippet for LSTM, as in, for example, NMT?

```python
tensor_info_x = utils.build_tensor_info(x)
tensor_info_y = utils.build_tensor_info(y)

prediction_signature = signature_def_utils.build_signature_def(
    inputs={'input': tensor_info_x},
    outputs={'output': tensor_info_y},
    method_name=signature_constants.PREDICT_METHOD_NAME)
```

**Comment by jinyuhang** [comment]
have you test your code in multi machine,such as 

python distribute_worker_mnist.py --ps_hosts=128.11.0.1:2222 --worker_hosts=128.12.0.1:2223,128.13.0.1:2224 --job_name=ps --task_index=0

python distribute_worker_mnist.py --ps_hosts=128.11.0.1:2222 --worker_hosts=128.12.0.1:2223,128.13.0.1:2224 --job_name=worker --task_index=0

python distribute_worker_mnist.py --ps_hosts=128.11.0.1:2222 --worker_hosts=128.12.0.1:2223,128.13.0.1:2224 --job_name=worker --task_index=1

**Comment by jinyuhang** [comment]
@terryKing1992 ,when i do so, it raise the error

W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at save_restore_v2_ops.cc:109 : Not found: ./dis_model/1/variables/variables_temp_ce6d1b67d4074ced87d9663082b8deb0; No such file or director
